### PR TITLE
Handle NIL in returned SQL result sets

### PR DESCRIPTION
**Pull request by Manuri**
## Purpose
Fix ballerina-platform/ballerina-lang#4024 
Fix ballerina-platform/ballerina-lang#7926

### CODE REVIEW

File: stdlib/ballerina-database/src/main/java/org/ballerinalang/database/sql/SQLDataIterator.java [hunk 1]
```diff
@@ -130,11 +135,19 @@ public BStruct generateNext() {
...
                     switch (sqlType) {
                     case Types.ARRAY:
                         Array dataArray = rs.getArray(index);
-                        bStruct.setRefField(++refRegIndex, getDataArray(dataArray));
+                        if (fieldTypeTag == TypeTags.UNION_TAG) {
+                            handleUnionTypeStructField(bStruct, structFields, index, ++refRegIndex,
+                                    getDataArray(dataArray), TypeTags.ARRAY_TAG);
+                        } else {
+                            bStruct.setRefField(++refRegIndex, getDataArray(dataArray));
+                        }
```
Comment (anupama-pathirage) [inline stdlib/ballerina-database/src/main/java/org/ballerinalang/database/sql/SQLDataIterator.java]: If all the values in the column are null in DB, then it will return an array of null values. So shouldn't we handle that as well?

File: stdlib/ballerina-database/src/main/java/org/ballerinalang/database/sql/SQLDataIterator.java [hunk 2]
```diff
@@ -255,14 +437,54 @@ public BStruct generateNext() {
...
         return bStruct;
     }
 
+    private void handleUnionTypeStructField(BStruct bStruct, BStructType.StructField[] structFields, int index,
+            int refRegIndex, BRefType value, int refType) {
+        if (isNillableType(structFields, index - 1, refType)) {
+            bStruct.setRefField(refRegIndex, value);
+        } else {
+            throw new BallerinaException("Union type contains more than 2 members or the members do not result in "
+                    + "an assignable nillable type");
```
Comment (anupama-pathirage) [inline stdlib/ballerina-database/src/main/java/org/ballerinalang/database/sql/SQLDataIterator.java]: Shall we rephrase this to  something like "Corresponding Union type in the record is not an assignable nillable type" as "Union type contains more than 2 members" implies any union type is ok if it has only two members. We only should give the idea that only nillable types are allowed. Also having index in the message is useful

File: stdlib/ballerina-database/src/main/java/org/ballerinalang/database/sql/SQLDataIterator.java [hunk 3]
```diff
@@ -130,11 +135,19 @@ public BStruct generateNext() {
                     SQLColumnDefinition def = (SQLColumnDefinition) columnDef;
                     columnName = def.getName();
                     int sqlType = def.getSqlType();
+                    BStructType.StructField[] structFields = this.type.getStructFields();
```
Comment (anupama-pathirage) [inline stdlib/ballerina-database/src/main/java/org/ballerinalang/database/sql/SQLDataIterator.java]: Shall we change this to `int fieldTypeTag = structFields[index - 1].getFieldType().getTag();`

File: stdlib/ballerina-database/src/main/java/org/ballerinalang/database/sql/SQLDataIterator.java [hunk 4]
```diff
@@ -143,107 +156,276 @@ public BStruct generateNext() {
...
-                        int fieldType = this.type.getStructFields()[index - 1].getFieldType().getTag();
-                        if (fieldType == TypeTags.STRING_TAG) {
-                            String dateValue = SQLDatasourceUtils.getString(date);
-                            bStruct.setStringField(++stringRegIndex, dateValue);
-                        } else if (fieldType == TypeTags.STRUCT_TAG) {
-                            bStruct.setRefField(++refRegIndex, createTimeStruct(date.getTime()));
-                        } else if (fieldType == TypeTags.INT_TAG) {
-                            bStruct.setIntField(++longRegIndex, date.getTime());
+                        if (fieldTypeTag == TypeTags.UNION_TAG) {
+                            if (isNillableType(structFields, index - 1, TypeTags.STRING_TAG)) {
```
Comment (anupama-pathirage) [inline stdlib/ballerina-database/src/main/java/org/ballerinalang/database/sql/SQLDataIterator.java]: Instead of checking the union type size and having nil in all if clauses, shall we break the logic in isNillableType method into two methods where,
1. check whether the union type size is 2 and one member is nil - call this before if-else-if ladder and throw an error.
2. check whether the member types contains the required type
This will eliminate duplicate checks while going through the cases.

File: stdlib/ballerina-database/src/main/java/org/ballerinalang/database/sql/SQLDataIterator.java [hunk 5]
```diff
@@ -143,107 +156,276 @@ public BStruct generateNext() {
...
-                        fieldType = this.type.getStructFields()[index - 1].getFieldType().getTag();
-                        if (fieldType == TypeTags.STRING_TAG) {
-                            String timeValue = SQLDatasourceUtils.getString(time);
-                            bStruct.setStringField(++stringRegIndex, timeValue);
-                        } else if (fieldType == TypeTags.STRUCT_TAG) {
-                            bStruct.setRefField(++refRegIndex, createTimeStruct(time.getTime()));
-                        } else if (fieldType == TypeTags.INT_TAG) {
-                            bStruct.setIntField(++longRegIndex, time.getTime());
+                        if (fieldTypeTag == TypeTags.UNION_TAG) {
+                            if (isNillableType(structFields, index - 1, TypeTags.STRING_TAG)) {
```
Comment (anupama-pathirage) [inline stdlib/ballerina-database/src/main/java/org/ballerinalang/database/sql/SQLDataIterator.java]: Refactor this too as per the above comment.

File: stdlib/ballerina-database/src/main/java/org/ballerinalang/database/sql/SQLDataIterator.java [hunk 6]
```diff
@@ -143,107 +156,276 @@ public BStruct generateNext() {
...
                     case Types.SMALLINT:
                         long iValue = rs.getInt(index);
-                        bStruct.setIntField(++longRegIndex, iValue);
+                        isOriginalValueNull = rs.wasNull();
+                        if (fieldTypeTag == TypeTags.UNION_TAG) {
+                            BRefType refValue = isOriginalValueNull ? null : new BInteger(iValue);
+                            handleUnionTypeStructField(bStruct, structFields, index, ++refRegIndex, refValue,
+                                    TypeTags.INT_TAG);
+                        } else {
+                            if (!isOriginalValueNull) {
```
Comment (anupama-pathirage) [inline stdlib/ballerina-database/src/main/java/org/ballerinalang/database/sql/SQLDataIterator.java]: What if we change the if else cases as follows by avoiding unnecessary ! here
if (isOriginalValueNull) {
handleNilToNonNilableFieldAssignment();
} else {
//set the value.
}

File: stdlib/ballerina-database/src/main/java/org/ballerinalang/database/sql/SQLDataIterator.java [hunk 7]
```diff
@@ -143,107 +156,276 @@ public BStruct generateNext() {
...
+                                handleNilToNonNilableFieldAssignment();
+                            }
+                        }
                         break;
                     case Types.STRUCT:
                         Struct structdata = (Struct) rs.getObject(index);
                         BType structFieldType = this.type.getStructFields()[index - 1].getFieldType();
-                        fieldType = structFieldType.getTag();
-                        if (fieldType == TypeTags.STRUCT_TAG) {
+                        if (fieldTypeTag == TypeTags.UNION_TAG) {
```
Comment (anupama-pathirage) [inline stdlib/ballerina-database/src/main/java/org/ballerinalang/database/sql/SQLDataIterator.java]: Shall we change this to BType structFieldType = structFields[index - 1].getFieldType();

File: stdlib/ballerina-database/src/main/java/org/ballerinalang/database/sql/SQLDataIterator.java [hunk 8]
```diff
@@ -255,14 +437,54 @@ public BStruct generateNext() {
...
+                    + "an assignable nillable type");
+        }
+    }
+
+    private void handleNilToNonNilableFieldAssignment() {
+        throw new BallerinaException("Trying to assign a Nil value to a non-nillable field");
+    }
+
+    private void handleMismatchingFieldAssignment() {
+        throw new BallerinaException("Trying to assign to a mismatching field");
```
Comment (anupama-pathirage) [inline stdlib/ballerina-database/src/main/java/org/ballerinalang/database/sql/SQLDataIterator.java]: Shall we add more information to this error message like field index and column name?

File: stdlib/ballerina-database/src/main/java/org/ballerinalang/database/sql/SQLDataIterator.java [hunk 9]
```diff
@@ -255,14 +437,54 @@ public BStruct generateNext() {
...
+        if (isNillableType(structFields, index - 1, refType)) {
+            bStruct.setRefField(refRegIndex, value);
+        } else {
+            throw new BallerinaException("Union type contains more than 2 members or the members do not result in "
+                    + "an assignable nillable type");
+        }
+    }
+
+    private void handleNilToNonNilableFieldAssignment() {
+        throw new BallerinaException("Trying to assign a Nil value to a non-nillable field");
```
Comment (anupama-pathirage) [inline stdlib/ballerina-database/src/main/java/org/ballerinalang/database/sql/SQLDataIterator.java]: Shall we add more information to this error message like field index and column name?

File: stdlib/ballerina-database/src/main/java/org/ballerinalang/database/sql/SQLDataIterator.java [hunk 10]
```diff
@@ -255,14 +437,54 @@ public BStruct generateNext() {
...
+    }
+
+    private void handleMismatchingFieldAssignment() {
+        throw new BallerinaException("Trying to assign to a mismatching field");
+    }
+
+    private boolean isNillableType(BStructType.StructField[] structFields, int index, int typeTag) {
+        List<BType> memberTypes = ((BUnionType) structFields[index].getFieldType())
+                .getMemberTypes();
```
Comment (anupama-pathirage) [inline stdlib/ballerina-database/src/main/java/org/ballerinalang/database/sql/SQLDataIterator.java]: Shall we reformat this to a single line?

File: stdlib/ballerina-database/src/main/java/org/ballerinalang/database/sql/SQLDataIterator.java [hunk 11]
```diff
@@ -255,14 +437,54 @@ public BStruct generateNext() {
...
+        if (memberTypes.size() != 2) {
+            return false;
+        }
+        boolean containsType = false;
+        boolean containsNil = false;
+        for (BType memberType : memberTypes) {
+            if (memberType.getTag() == TypeTags.UNION_TAG) {
+                return false;
```
Comment (anupama-pathirage) [inline stdlib/ballerina-database/src/main/java/org/ballerinalang/database/sql/SQLDataIterator.java]: Why do we need to check union types here?

File: stdlib/ballerina-database/src/main/java/org/ballerinalang/database/sql/SQLDataIterator.java [hunk 12]
```diff
@@ -255,14 +437,54 @@ public BStruct generateNext() {
...
+            }
+            if (memberType.getTag() == typeTag) {
```
Comment (anupama-pathirage) [inline stdlib/ballerina-database/src/main/java/org/ballerinalang/database/sql/SQLDataIterator.java]: Shall we assign the tag value to an int at the top of the loop and then do checks?

File: tests/ballerina-test/src/test/java/org/ballerinalang/test/connectors/sql/SQLActionsTest.java [hunk 13]
```diff
@@ -648,6 +654,416 @@ public void testLoadToMemorySelectAfterTableClose() throws Exception {
...
+        Assert.assertNull(returns[14]);
+        Assert.assertNull(returns[15]);
+        Assert.assertNull(returns[16]);
+        Assert.assertNull(returns[17]);
+    }
+
+    @Test(groups = "ConnectorTest",
+          description = "Test mapping nil to non-nillable int field",
+          expectedExceptions = BLangRuntimeException.class,
+          expectedExceptionsMessageRegExp = ".*" + TRYING_TO_ASSIGN_NIL_TO_NON_NILLABLE_FIELD + ".*")
```
Comment (anupama-pathirage) [inline tests/ballerina-test/src/test/java/org/ballerinalang/test/connectors/sql/SQLActionsTest.java]: Instead of having concat at the beginning and end shall we define a constant with those two parts as well

File: tests/ballerina-test/src/test/resources/test-src/connectors/sql/sql_actions_test.bal [hunk 14]
```diff
@@ -60,12 +89,38 @@ type ResultBalTypes {
...
+    time:Time? timestamp_type,
+};
+
 type Employee {
     int id,
     string name,
     string address,
 };
 
+type Employee1 {
```
Comment (anupama-pathirage) [inline tests/ballerina-test/src/test/resources/test-src/connectors/sql/sql_actions_test.bal]: Unused type?

File: tests/ballerina-test/src/test/resources/test-src/connectors/sql/sql_actions_test.bal [hunk 15]
```diff
@@ -1176,6 +1231,229 @@ function testLoadToMemorySelectAfterTableClose() returns (Employee[], Employee[]
...
+        float?, int?, int?, string?, blob?, blob?, time:Time?, time:Time?, time:Time?, time:Time?) {
+    endpoint jdbc:Client testDB {
+        url: "jdbc:hsqldb:file:./target/tempdb/TEST_SQL_CONNECTOR",
+        username: "SA",
+        poolOptions: { maximumPoolSize: 1 }
+    };
+    table<NillableDataTypes> dt = check testDB->select("SELECT int_type, long_type, float_type, double_type,
+    boolean_type, string_type,
+  numeric_type, decimal_type, real_type, tinyint_type, smallint_type, clob_type, blob_type, binary_type, date_type,
+  time_type, datetime_type, timestamp_type from DataTypeTableNillable where row_id=1", NillableDataTypes);
```
Comment (anupama-pathirage) [inline tests/ballerina-test/src/test/resources/test-src/connectors/sql/sql_actions_test.bal]: Reformat the code to fix the alignment

File: tests/ballerina-test/src/test/resources/test-src/connectors/sql/sql_actions_test.bal [hunk 16]
```diff
@@ -1176,6 +1231,229 @@ function testLoadToMemorySelectAfterTableClose() returns (Employee[], Employee[]
...
+        binary_type = rs.binary_type;
+        date_type = rs.date_type;
+        time_type = rs.time_type;
+        datetime_type = rs.datetime_type;
+        timestamp_type = rs.timestamp_type;
+    }
+    testDB.stop();
+    return (int_type, long_type, float_type, double_type,
+                boolean_type, string_type,
+              numeric_type, decimal_type, real_type, tinyint_type, smallint_type, clob_type, blob_type, binary_type, date_type,
```
Comment (anupama-pathirage) [inline tests/ballerina-test/src/test/resources/test-src/connectors/sql/sql_actions_test.bal]: Refomat the code to avoid more than 120 chars

File: tests/ballerina-test/src/test/resources/test-src/connectors/sql/sql_actions_test.bal [hunk 17]
```diff
@@ -1176,6 +1231,229 @@ function testLoadToMemorySelectAfterTableClose() returns (Employee[], Employee[]
...
+            float?, int?, int?, string?, blob?, blob?, time:Time?, time:Time?, time:Time?, time:Time?) {
+    endpoint jdbc:Client testDB {
+        url: "jdbc:hsqldb:file:./target/tempdb/TEST_SQL_CONNECTOR",
+        username: "SA",
+        poolOptions: { maximumPoolSize: 1 }
+    };
+    table<NillableDataTypes> dt = check testDB->select("SELECT int_type, long_type, float_type, double_type,
+    boolean_type, string_type,
+  numeric_type, decimal_type, real_type, tinyint_type, smallint_type, clob_type, blob_type, binary_type, date_type,
+  time_type, datetime_type, timestamp_type from DataTypeTableNillable where row_id=2", NillableDataTypes);
```
Comment (anupama-pathirage) [inline tests/ballerina-test/src/test/resources/test-src/connectors/sql/sql_actions_test.bal]: Reformat the line

File: tests/ballerina-test/src/test/resources/test-src/connectors/sql/sql_actions_test.bal [hunk 18]
```diff
@@ -1176,6 +1231,229 @@ function testLoadToMemorySelectAfterTableClose() returns (Employee[], Employee[]
...
+        date_type = rs.date_type;
+        time_type = rs.time_type;
+        datetime_type = rs.datetime_type;
+        timestamp_type = rs.timestamp_type;
+    }
+    testDB.stop();
+    return (int_type, long_type, float_type, double_type,
+    boolean_type, string_type,
+    numeric_type, decimal_type, real_type, tinyint_type, smallint_type, clob_type, blob_type, binary_type, date_type,
+    time_type, datetime_type, timestamp_type);
```
Comment (anupama-pathirage) [inline tests/ballerina-test/src/test/resources/test-src/connectors/sql/sql_actions_test.bal]: Reformat the line, may be we can fit it into 3 lines only

File: tests/ballerina-test/src/test/resources/test-src/connectors/sql/sql_actions_negative_test.bal [hunk 19]
```diff
@@ -16,6 +17,82 @@ type ResultCustomers2 {
     string LASTNAME,
 };
 
+type NonNillableInt {
```
Comment (anupama-pathirage) [inline tests/ballerina-test/src/test/resources/test-src/connectors/sql/sql_actions_negative_test.bal]: Seems all the types other than InvalidUnion1 and NonNillableInt is not used.

File: tests/ballerina-test/src/test/resources/test-src/connectors/sql/sql_actions_negative_test.bal [hunk 20]
```diff
@@ -16,6 +17,82 @@ type ResultCustomers2 {
...
+
+type NonNillableDateTime {
+    time:Time val;
+};
+
+type NonNillableTimeStamp {
+    time:Time val;
+};
+
+type InvalidUnion1 {
```
Comment (anupama-pathirage) [inline tests/ballerina-test/src/test/resources/test-src/connectors/sql/sql_actions_negative_test.bal]: Shall we rename this to InvalidUnion

**Comment by anupama-pathirage** [comment]
@Manuri It seems some tests are failing, see the travis log.

File: tests/ballerina-test/src/test/resources/test-src/connectors/sql/sql_actions_negative_test.bal [hunk 21]
```diff
@@ -1,6 +1,7 @@
 import ballerina/sql;
 import ballerina/jdbc;
 import ballerina/io;
+import ballerina/time;
```
Comment (anupama-pathirage) [inline tests/ballerina-test/src/test/resources/test-src/connectors/sql/sql_actions_negative_test.bal]: Unused import?

File: tests/ballerina-test/src/test/resources/test-src/connectors/sql/sql_actions_negative_test.bal [hunk 22]
```diff
@@ -273,3 +350,185 @@ function testCallProcedureWithMultipleResultSetsAndNilConstraintCount() returns
...
+
+function testAssignToInvalidUnionDateTime() {
+    testAssignToInvalidUnionField("datetime_type");
+}
+
+function testAssignToInvalidUnionTimeStamp() {
+    testAssignToInvalidUnionField("timestamp_type");
+}
+
+
```
Comment (anupama-pathirage) [inline tests/ballerina-test/src/test/resources/test-src/connectors/sql/sql_actions_negative_test.bal]: Extra new line

File: tests/ballerina-test/src/test/resources/test-src/connectors/sql/sql_actions_negative_test.bal [hunk 23]
```diff
@@ -273,3 +350,185 @@ function testCallProcedureWithMultipleResultSetsAndNilConstraintCount() returns
...
+    try {
+        while (dt.hasNext()) {
+            var rs = <InvalidUnion1>dt.getNext();
+        }
+    } finally {
+        testDB.stop();
+    }
+}
+
```
Comment (anupama-pathirage) [inline tests/ballerina-test/src/test/resources/test-src/connectors/sql/sql_actions_negative_test.bal]: Extra new line

File: tests/ballerina-test/src/test/java/org/ballerinalang/test/connectors/sql/SQLActionsTest.java [hunk 24]
```diff
@@ -648,6 +654,416 @@ public void testLoadToMemorySelectAfterTableClose() throws Exception {
...
+        double realExpected = 1234.567;
+        long tinyIntExpected = 1;
+        long smallIntExpected = 5555;
+        String clobExpected = "very long text";
+        String blobExpected = "wso2 ballerina blob test.";
+        String binaryExpected = "wso2 ballerina binary test.";
+        String dateExpected = "{time:1486060200000, zone:{zoneId:\"UTC\", zoneOffset:0}}";
+        String dateTimeExpected = "{time:41745000, zone:{zoneId:\"UTC\", zoneOffset:0}}";
+        String timeExpected = "{time:1486122780000, zone:{zoneId:\"UTC\", zoneOffset:0}}";
+        String timeStampExpected = "{time:1486122780000, zone:{zoneId:\"UTC\", zoneOffset:0}}";
```
Comment (anupama-pathirage) [inline tests/ballerina-test/src/test/java/org/ballerinalang/test/connectors/sql/SQLActionsTest.java]: Shall we directly check for the expected values rather than defining separate variables, It will reduce the code a lot and improves readability

File: stdlib/ballerina-database/src/main/java/org/ballerinalang/database/sql/SQLDataIterator.java (hunk 2 above)
Comment (Manuri) [reply stdlib/ballerina-database/src/main/java/org/ballerinalang/database/sql/SQLDataIterator.java]: Will re-phrase the message. The index is automatically added by a higher level exception message so I did not add it here.

File: stdlib/ballerina-database/src/main/java/org/ballerinalang/database/sql/SQLDataIterator.java (hunk 8 above)
Comment (Manuri) [reply stdlib/ballerina-database/src/main/java/org/ballerinalang/database/sql/SQLDataIterator.java]: Index is already being added at a higher level. I think we can add the column name.

File: stdlib/ballerina-database/src/main/java/org/ballerinalang/database/sql/SQLDataIterator.java (hunk 11 above)
Comment (anupama-pathirage) [reply stdlib/ballerina-database/src/main/java/org/ballerinalang/database/sql/SQLDataIterator.java]: But union type never get passed into this as typeTag and even if the ballerina code has a union type it will get validated from the below if case. So we don't need this.

File: tests/ballerina-test/src/test/resources/test-src/connectors/sql/sql_actions_negative_test.bal (hunk 21 above)
Comment (Manuri) [reply tests/ballerina-test/src/test/resources/test-src/connectors/sql/sql_actions_negative_test.bal]: This is used in several types.

File: stdlib/ballerina-database/src/main/java/org/ballerinalang/database/sql/SQLDataIterator.java [hunk 25]
```diff
@@ -325,6 +311,261 @@ private BStruct createUserDefinedType(Struct structValue, BStructType structType
...
+                handleNilToNonNillableFieldAssignment();
+            } else {
+                bStruct.setFloatField(doubleRegIndex.incrementAndGet(), fValue);
+            }
+        }
+    }
+
+    private void handleMappingDateValueToUnionType(BStructType.StructField[] structFields, int index, BStruct bStruct,
+            RegistryIndex refRegIndex, java.util.Date date) {
+        int type = retrieveType(structFields, index - 1);
```
Comment (anupama-pathirage) [inline stdlib/ballerina-database/src/main/java/org/ballerinalang/database/sql/SQLDataIterator.java]: Don't we need to validate using validateAndSetRefRecordField method?

File: stdlib/ballerina-database/src/main/java/org/ballerinalang/database/sql/SQLDataIterator.java [hunk 26]
```diff
@@ -325,6 +311,261 @@ private BStruct createUserDefinedType(Struct structValue, BStructType structType
...
+            bStruct.setRefField(refRegIndex.incrementAndGet(), date != null ? new BInteger(date.getTime()) : null);
+            break;
+        default:
+            handleMismatchingFieldAssignment();
+        }
+    }
+
+    private void handleMappingDateValueToNonUnionType(java.util.Date date, int fieldTypeTag, BStruct bStruct,
+            RegistryIndex stringRegIndex, RegistryIndex refRegIndex, RegistryIndex longRegIndex) {
+        if (date != null) {
```
Comment (anupama-pathirage) [inline stdlib/ballerina-database/src/main/java/org/ballerinalang/database/sql/SQLDataIterator.java]: Don't we need to validate using validateAndSetRefRecordField method?

File: stdlib/ballerina-database/src/main/java/org/ballerinalang/database/sql/SQLDataIterator.java [hunk 27]
```diff
@@ -325,6 +311,261 @@ private BStruct createUserDefinedType(Struct structValue, BStructType structType
...
+        }
+        handleDoubleValue(decimalValue, bStruct, refRegIndex, doubleRegIndex, index);
+    }
+
+    private void handleDoubleValue(double fValue, BStruct bStruct, RegistryIndex refRegIndex,
+            RegistryIndex doubleRegIndex, int index) throws SQLException {
+        BStructType.StructField[] structFields = getStructFields();
+        boolean isOriginalValueNull = rs.wasNull();
+        int fieldTypeTag = getFieldTypeTag(structFields, index);
+
```
Comment (anupama-pathirage) [inline stdlib/ballerina-database/src/main/java/org/ballerinalang/database/sql/SQLDataIterator.java]: Extra new line?

File: tests/ballerina-test/src/test/java/org/ballerinalang/test/types/table/TableTest.java [hunk 28]
```diff
@@ -50,11 +51,22 @@
...
+            + "assignable nillable type.*";
+    private static final double DELTA = 0.01;
 
     @BeforeClass
     public void setup() {
         result = BCompileUtil.compile("test-src/types/table/table_type.bal");
+        nillableMappingNegativeResult = BCompileUtil.compile("test-src/types/table/table_nillable_mapping_negative"
+                + ".bal");
+        nillableMappingResult = BCompileUtil.compile
+                ("test-src/types/table/table_nillable_mapping.bal");
```
Comment (anupama-pathirage) [inline tests/ballerina-test/src/test/java/org/ballerinalang/test/types/table/TableTest.java]: This can be a single line

File: tests/ballerina-test/src/test/java/org/ballerinalang/test/types/table/TableTest.java [hunk 29]
```diff
@@ -583,8 +564,436 @@ public void testTableRemoveInvalid() {
...
+    @Test(groups = "TableTest",
+          description = "Test mapping date to nillable Time field")
+    public void testMapptingDatesToNillableTime() {
+        BValue[] returns = BRunUtil.invoke(nillableMappingResult,  "testMappingDatesToNillableTimeType");
+        Assert.assertEquals(returns.length, 8);
+        Assert.assertEquals(((BInteger) returns[0]).intValue(), ((BInteger) returns[1]).intValue());
+        Assert.assertEquals(((BInteger) returns[2]).intValue(), ((BInteger) returns[3]).intValue());
+        Assert.assertEquals(((BInteger) returns[4]).intValue(), ((BInteger) returns[5]).intValue());
+        Assert.assertEquals(((BInteger) returns[6]).intValue(), ((BInteger) returns[7]).intValue());
+
```
Comment (anupama-pathirage) [inline tests/ballerina-test/src/test/java/org/ballerinalang/test/types/table/TableTest.java]: Extra new line?

File: tests/ballerina-test/src/test/java/org/ballerinalang/test/utils/SQLDBUtils.java [hunk 30]
```diff
@@ -63,7 +66,7 @@ public static void initDatabase(String dbDirectory, String dbName, String sqlFil
                 st.executeUpdate(query.trim());
             }
         } catch (ClassNotFoundException | SQLException e) {
-            //Do nothing
+            log.error("Error while initializing database: ", e);
```
Comment (anupama-pathirage) [inline tests/ballerina-test/src/test/java/org/ballerinalang/test/utils/SQLDBUtils.java]: Do the same for initH2Database method as well

File: tests/ballerina-test/src/test/resources/test-src/types/table/table_nillable_mapping.bal [hunk 31]
```diff
@@ -0,0 +1,310 @@
...
+        smallint_type = rs.smallint_type;
+        clob_type = rs.clob_type;
+        blob_type = rs.blob_type;
+        binary_type = rs.binary_type;
+    }
+    testDB.stop();
+    return (int_type, long_type, float_type, double_type,
+    boolean_type, string_type,
+    numeric_type, decimal_type, real_type, tinyint_type, smallint_type, clob_type, blob_type, binary_type);
+
```
Comment (anupama-pathirage) [inline tests/ballerina-test/src/test/resources/test-src/types/table/table_nillable_mapping.bal]: Extra new line

File: tests/ballerina-test/src/test/resources/test-src/types/table/table_nillable_mapping.bal [hunk 32]
```diff
@@ -0,0 +1,310 @@
...
+        date_type = rs.date_type;
+        time_type = rs.time_type;
+        datetime_type = rs.datetime_type;
+        timestamp_type = rs.timestamp_type;
+    }
+    testDB.stop();
+    return (int_type, long_type, float_type, double_type,
+    boolean_type, string_type,
+    numeric_type, decimal_type, real_type, tinyint_type, smallint_type, clob_type, blob_type, binary_type, date_type,
+    time_type, datetime_type, timestamp_type);
```
Comment (anupama-pathirage) [inline tests/ballerina-test/src/test/resources/test-src/types/table/table_nillable_mapping.bal]: Reformat the line. Should be able to fit to 3 lines

File: tests/ballerina-test/src/test/resources/test-src/types/table/table_nillable_mapping.bal [hunk 33]
```diff
@@ -0,0 +1,310 @@
...
+        float?, int?, int?, string?, blob?, blob?, time:Time?, time:Time?, time:Time?, time:Time?) {
+    endpoint jdbc:Client testDB {
+        url: "jdbc:hsqldb:file:./target/tempdb/TEST_DATA_TABLE_DB",
+        username: "SA",
+        poolOptions: { maximumPoolSize: 1 }
+    };
+    table<NillableDataTypes> dt = check testDB->select("SELECT int_type, long_type, float_type, double_type,
+    boolean_type, string_type,
+  numeric_type, decimal_type, real_type, tinyint_type, smallint_type, clob_type, blob_type, binary_type, date_type,
+  time_type, datetime_type, timestamp_type from DataTypeTableNillable where row_id=2", NillableDataTypes);
```
Comment (anupama-pathirage) [inline tests/ballerina-test/src/test/resources/test-src/types/table/table_nillable_mapping.bal]: Formatting issue

File: tests/ballerina-test/src/test/resources/test-src/types/table/table_nillable_mapping.bal [hunk 34]
```diff
@@ -0,0 +1,310 @@
...
+        time = rs.TIME_TYPE but { () => -1 };
+        date = rs.DATE_TYPE but { () => -1 };
+        timestamp = rs.TIMESTAMP_TYPE but { () => -1 };
+        datetime = rs.DATETIME_TYPE but { () => -1 };
+    }
+    testDB.stop();
+    return (date, time, timestamp, datetime);
+}
+
+function testMappingDatesToNillableStringType(int datein, int timein, int timestampin) returns (string, string, string, string) {
```
Comment (anupama-pathirage) [inline tests/ballerina-test/src/test/resources/test-src/types/table/table_nillable_mapping.bal]: Reformat

File: tests/ballerina-test/src/test/resources/test-src/types/table/table_nillable_mapping.bal [hunk 35]
```diff
@@ -0,0 +1,310 @@
...
+    int time;
+    int timestamp;
+    int datetime;
+
+    int countt = check testDB->update("Insert into DateTimeTypes
+        (row_id, date_type, time_type, timestamp_type, datetime_type) values (?,?,?,?,?)",
+        para0, para1, para2, para3, para4);
+
+    table<ResultDatesWithNillableIntType> dt = check testDB->select("SELECT date_type, time_type, timestamp_type, datetime_type
+                from DateTimeTypes where row_id = 151", ResultDatesWithNillableIntType);
```
Comment (anupama-pathirage) [inline tests/ballerina-test/src/test/resources/test-src/types/table/table_nillable_mapping.bal]: Reformat the line

File: tests/ballerina-test/src/test/resources/test-src/types/table/table_nillable_mapping_negative.bal [hunk 36]
```diff
@@ -0,0 +1,275 @@
...
+type NonNillableDateTime {
+    time:Time val;
+};
+
+type NonNillableTimeStamp {
+    time:Time val;
+};
+
+type InvalidUnion {
+    (int | string) val;
```
Comment (anupama-pathirage) [inline tests/ballerina-test/src/test/resources/test-src/types/table/table_nillable_mapping_negative.bal]: Formatting issue

File: tests/ballerina-test/src/test/resources/test-src/types/table/table_nillable_mapping_negative.bal [hunk 37]
```diff
@@ -0,0 +1,275 @@
...
+};
+
+
```
Comment (anupama-pathirage) [inline tests/ballerina-test/src/test/resources/test-src/types/table/table_nillable_mapping_negative.bal]: Extra new line

File: stdlib/ballerina-database/src/main/java/org/ballerinalang/database/sql/SQLDataIterator.java (hunk 25 above)
Comment (Manuri) [reply stdlib/ballerina-database/src/main/java/org/ballerinalang/database/sql/SQLDataIterator.java]: For Date, we have several possible assignable field types "string", "struct" and "int". So there is a switch below this line of which the default case is throwing an exception. So the validation happens. 

validateAndSetRefRecordField is used for other types which have only one field type.

File: stdlib/ballerina-database/src/main/java/org/ballerinalang/database/sql/SQLDataIterator.java (hunk 26 above)
Comment (Manuri) [reply stdlib/ballerina-database/src/main/java/org/ballerinalang/database/sql/SQLDataIterator.java]: For Date, we have several possible assignable field types "string", "struct" and "int". So there is a switch below this line of which the default case is throwing an exception. So the validation happens. 

validateAndSetRefRecordField is used for other types which have only one field type.
